lib/ClangImporter/ImportDecl.cpp

```diff
@@ -7579,25 +7579,30 @@ synthesizeSubscriptGetterBody(AbstractFunctionDecl *afd, void *context) {
                                                         inoutSelfExpr,
                                                         keyRefExpr);
 
-  // `getterImpl` can return either UnsafePointer or UnsafeMutablePointer.
-  // Retrieve the corresponding `.pointee` declaration.
+  // This default handles C++'s operator[] that returns a value type.
+  Expr *propertyExpr = getterImplCallExpr;
   PointerTypeKind ptrKind;
-  getterImpl->getResultInterfaceType()->getAnyPointerElementType(ptrKind);
-  VarDecl *pointeePropertyDecl = ctx.getPointerPointeePropertyDecl(ptrKind);
-
-  SubstitutionMap subMap =
-      SubstitutionMap::get(ctx.getUnsafePointerDecl()->getGenericSignature(),
-                           { elementTy }, { });
-  auto pointeePropertyRefExpr =
-      new (ctx) MemberRefExpr(getterImplCallExpr,
-                              SourceLoc(),
-                              ConcreteDeclRef(pointeePropertyDecl, subMap),
-                              DeclNameLoc(),
-                              /*implicit=*/ true);
-  pointeePropertyRefExpr->setType(elementTy);
+  if (getterImpl->getResultInterfaceType()->getAnyPointerElementType(ptrKind)) {
```

Comment:
Is this path also going to be taken for a pointer overload (specifically a pointer type, not a reference type)?

Reply:
Let me get back to you on this. It definitely was taken for reference types.

Reply:
Yep, this block only gets invoked for the reference types.

Reply:
@zoecarver You want this to only be invoked by the reference type path correct?

Reply:
I see, so the reason this branch path only applies for reference types is, the reference types are a `UnsafeMutablePointer<Int32>` / `UnsafePointer<Int32>` whereas the pointer types are Optionals so they are `UnsafeMutablePointer<Int32>?` / `UnsafePointer<Int32>?`. `getAnyPointerElementType` returns a `Type()` for these since they aren't any of the five `PointerTypeKind`s (ie   `PTK_UnsafeMutableRawPointer`, `PTK_UnsafeRawPointer`, `PTK_UnsafeMutablePointer`, `PTK_UnsafePointer`, `PTK_AutoreleasingUnsafeMutablePointer`).

Reply:
I decided to go with a comment and a todo to use a better conditional check. Is that ok for now?

Reply:
I'm going to mark this thread resolved. Un-resolve it if you do feel strongly.